File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042839. 소수 찾기/kotlin/Solution.kt
import kotlin.math.sqrt

class Solution {
    fun solution(numbers: String): Int {
        val max = (1).let{
            var result = it
            repeat(numbers.length) { result *= 10 }
            result
        }
        val isPrime = BooleanArray(max) {true}
        isPrime[0] = false
        isPrime[1] = false
        for (i in 2..sqrt(max.toDouble()).toInt()) {
            if (isPrime[i]) {
                for (j in 2*i until max step i) {
                    isPrime[j] = false
                }
            }
        }
        val set = HashSet<Int>()
        val check = BooleanArray(numbers.length)
        fun dfs(v: Int, cnt: Int) {
            var next: Int
            for (i in numbers.indices) {
                if (!check[i]) {
                    check[i] = true
                    next = v * 10 + numbers[i].code - 48
                    if (isPrime[next]) set += next
                    if (cnt + 1 < numbers.length) dfs(next, cnt + 1)
                    check[i] = false
                }
            }
        }
        dfs(0, 0)
        return set.size
    }


}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042839. 소수 찾기/python/solution.py
from math import sqrt


def solution(numbers):
    s = set()
    check = [False] * len(numbers)
    max_num = -1

    def dfs(idx, value):
        nonlocal max_num, check
        for i in range(len(numbers)):
            if idx == 0 and numbers[i] == '0':
                continue
            if not check[i]:
                check[i] = True
                next_value = value * 10 + ord(numbers[i]) - ord('0')
                if next_value > max_num:
                    max_num = next_value
                s.add(next_value)
                if idx != len(numbers) - 1:
                    dfs(idx + 1, next_value)
                check[i] = False

    dfs(0, 0)

    is_prime_number = [True] * (max_num + 1)
    is_prime_number[0], is_prime_number[1] = False, False
    if 1 in s:
        s.remove(1)
    for i in range(2, int(sqrt(max_num)) + 1):
        if is_prime_number[i]:
            for j in range(2 * i, max_num + 1, i):
                is_prime_number[j] = True
                if j in s:
                    s.remove(j)
    return len(s)
